File: codebreakers/vigenere.rb
class Vigenere
  attr_reader :keys

  def letters
    ('a'..'z').to_a
  end

  def initialize
    @keys = get_indices
  end

  def get_indices
    keyword = ARGV[0].downcase
    keyword.chars.map do |c|
      letters.index(c)+1
    end
  end

  def encode
    message = ARGV[1..-1].join(" ")
    letters_count = 0
    encoded = message.chars.map do |c|
      ind = letters.index(c.downcase)
      new_ind = (ind + keys[letters_count % keys.length]) % 26 if ind
      letters_count += 1 if ind
      if ind && c == c.upcase
        letters[new_ind].upcase
      elsif ind && c == c.downcase
        letters[new_ind]
      else
        c
      end

    end
    puts encoded.join
  end


end

if __FILE__ == $0
  vig = Vigenere.new
  vig.encode
end

File: codebreakers/vigenere.py
from sys import argv
from string import uppercase, lowercase

def main():
    keyword = argv[1]
    message = " ".join(argv[2:])
    keys = [lowercase.index(c.lower())+1 for c in keyword]
    letter_count = 0
    encoded = []
    for c in message:
        if c.lower() in lowercase:
            ind = lowercase.index(c.lower())
            new_ind = (ind + keys[letter_count % len(keys)]) % 26
            letter_count += 1
        if c in uppercase:
            encoded.append(uppercase[new_ind])
        elif c in lowercase:
            encoded.append(lowercase[new_ind])
        else:
            encoded.append(c)
    print "".join(encoded)

if __name__ == '__main__':
    main()
